### ISSUE TITLE
Chainstate corrupted on normal exit (Linux 18.04 64-bit) -- PublicCoin::validate value is not prime

### ISSUE THREAD

**Issue by seanPhill** [Veil-Project/veil#371]

**Comment by seanPhill** [comment]
I closed the (mainnet) wallet normally and completed some Linux software updates. Restarted the computer (VM) and on starting the wallet got an error with chainstate (twice)! I deleted the chainstate folder and started again. The wallet proceeded to rebuild the chainstate, but when I  got back to check it the wallet was closed/crashed again, with an error at the command line about the validation value not being prime!

```$ ./veil-qt
veil-qt: validation.cpp:3265: void CChainState::PruneBlockIndexCandidates(): Assertion `!setBlockIndexCandidates.empty()' failed.
Aborted (core dumped)
vagrant@linux:~/veil$ ./veil-qt
veil-qt: validation.cpp:3265: void CChainState::PruneBlockIndexCandidates(): Assertion `!setBlockIndexCandidates.empty()' failed.
Aborted (core dumped)
```
```
$ ./veil-qt
PublicCoin::validate value is not prime
```
Launched the wallet again and it is checking the blocks from January **4th** again.

**Comment by seanPhill** [comment]
The current issue, post-deletion of Chainstate and retrying matches "CheckZerocoinMint() : PubCoin does not validate - Occasionally Generated PublicCoin Becomes Mutated #230"

**Comment by seanPhill** [comment]
Eight hours later, after rebuilding the chainstate, I got "a fatal internal error" "see the debug log for details."
[LinuxDebugLog(portion).txt](https://github.com/Veil-Project/veil/files/2842244/LinuxDebugLog.portion.txt)

**Comment by presstab** [comment]
The `PublicCoin::validate` is not a chainstate issue, it is a tx stored in your wallet that is invalid.

The issue here appears to be this:
```
2019-02-07T18:01:20Z ERROR: ReadBlockFromDisk(CBlock&, CBlockIndex*): GetHash() doesn't match index for CBlockIndex(pprev=0x7f74e0997490, nHeight=49505, merkle=0000000000000000000000000000000000000000000000000000000000000000, hashBlock=4f36ddefc752527dc154e79d2978df39b8de621b4de9906100007f74e0993180) at CBlockDiskPos(nFile=14, nPos=99330759)
```

**Comment by presstab** [comment]
Deleting the chainstate folder itself is probably not the correct way to do things. Would maybe want to restart with `-reindex` or `-reindex-chainstate`

**Comment by seanPhill** [comment]
Thanks. It's currently syncing starting from Feb 3rd snapshot.

**Comment by seanPhill** [comment]
At the end of the sync I got **again** ```Error: A fatal internal error occurred, see debug.log for details```

**Comment by seanPhill** [comment]
I think this will want more than a reindex.

**Comment by seanPhill** [comment]
I restored from seed words with just the blocks, chainstate and zerocoin folders remaining. Crashed again pretty quickly.
```
vagrant@linux:~/veil$ ./veil-qt
Segmentation fault (core dumped)
vagrant@linux:~/veil$ 
```

**Comment by seanPhill** [comment]
Just finished syncing from the snapshot!
I just got 
```Error: A fatal internal error occurred, see debug.log for details```
 AGAIN! This is with a fresh wallet from seed words and resync from the snapshot. It's taken since about 5 AM. 9 PM now.

**Comment by seanPhill** [comment]
This one again! (All those files were deleted, and replaced by the snapshot plus syncing.) 
```
2019-02-08T01:39:33Z ERROR: ReadBlockFromDisk(CBlock&, CBlockIndex*): GetHash() doesn't match index for CBlockIndex(pprev=0x7f0993f64de0, nHeight=49505, merkle=0000000000000000000000000000000000000000000000000000000000000000, hashBlock=4f36ddefc752527dc154e79d2978df39b8de621b4de9906100007f0993f60ad0) at CBlockDiskPos(nFile=14, nPos=99330759)
```

**Comment by seanPhill** [comment]
[debug-portion.txt](https://github.com/Veil-Project/veil/files/2843375/debug-portion.txt)

**Comment by seanPhill** [comment]
Just realised that the block that crashed the wallet, **because of an index mismatch** has two very suspicious things about it!!

1. It is the same block height as this one: 
4f36ddefc752527dc154e79d2978df39b8de621b4de990614b685e487cbaecb4
which has a received block **mining** reward of ```49.93145424 Veil``` !!!
2. Its hash is astronomically unlikely to be legitimately generated!!!!!!!
4f36ddefc752527dc154e79d2978df39b8de621b4de9906100007f0993f60ad0

Look how similar that is!!

My wallet synced that block and crashed TWICE, after loading a snapshot (that does not contain that block height -- it synced afterwards). That block 49505 is different to the one on the block explorer, but I wonder by how much! Maybe just the difference between a mining reward of 50.00000000 and 49.93145424?? (Or maybe that plus a different winning wallet address????)

**Comment by seanPhill** [comment]
I have abandoned this wallet in Linux (for now) and have restored from seed words on macOS. Just got to regenerate the right addresses to get my correct balance. 
```restoreaddresses 10``` (Now on the second ten.)
```rescanblockchain``` (Again)
```searchdeterministiczerocoin``` (1000 done)

1000 Veil still missing. (Maybe try the deterministic search again.) Did the same 1000. 
And found the remaining zerocoin.

In the above process, after I got the blockchain fully synced and not crashing I also tested replacing back the wallet.dat I had before, but that one crashed again, every time. so the wallet.dat itself had become corrupted. So I put the clean wallet.dat back in and did the above steps.
